### ISSUE TITLE
Case for unknowns

### ISSUE THREAD

**Issue by hariszaf** [msysbio/CellScanner#11]
@hallucigenia-sparsa we were discussing with @ccvdvelde today, and she realized we are missing the `Unknown` cases in this CellScanner version.
It seems this can be quite handy so you can have an idea on how bad a dataset is just by looking the number of `Unknown`s you have. 

The [`tensorflow.keras.model.predict()`](https://github.com/keras-team/keras/blob/f6c4ac55692c132cd16211f4877fac6dbeead749/keras/src/backend/tensorflow/trainer.py#L438C10-L514C1) way we're using now to get the predictions does not support such a feature on its own. 

Karoline, we were thinking with Charlotte whether we could use a threshold for the uncertainty and assign as `Unknown` anything with an uncertainty higher than that. What do you think ?

**Comment by hariszaf** [comment]
Ok, apologies if I get the concepts a bit slow. 😅 

I was able to figure out that the maximum value of the uncertainty score for an event, it would be equal to the `ln(number_of_classes)`; log with base `e`. For example, if we have `blank`, `BT`, `RI` and `FP`, we would have 4 classes, thus maximum uncertainty of `ln(4)`. 

I know now what entropy and thus uncertainties stand for, but I am still trying to figure out how to come up with a threshold. I ll text you here for any news.

**Comment by hallucigenia-sparsa** [comment]
sorry, should have clarified better (see [here](http://msysbiology.com/documents/CellScanner/Differences_CellScanner_1_and_2.pdf) for comparison of CS1 and CS2). True, entropy is not bounded, so cannot threshold on percentage as we could in CellScanner 1 (which used the percentage of classifiers that had to agree on the label).

**Comment by hariszaf** [comment]
Thanks for the link! 

Based on what I have found online, here are some thoughts:

- (easiest) it seems 0.5 is a default value, that's why (I guess) Ermis had that in the file produced
- (easy way) percentile-Based Threshold: the user chooses a percentile as the cutoff for uncertainty; the user in this case provides the percentile and the threshold is computed to only keep the percentage of points asked -- i do not like it since it cannot inform us about the quality of the data
- (a bit harder) Accuracy-Based Threshold (requires validation datasets): pick a set of entropy thresholds and check how they affect classification accuracy. Keep the threshold where performance starts to drop significantly.

**Comment by hallucigenia-sparsa** [comment]
if fast enough, the last option is the best. Since classification accuracy is always computed with the monocultures (in-silico mixtures), this option doesn't require validation data sets (though they would be better). So uncertainty thresholding is an option - if selected, the threshold can be optionally provided by the user and if not, it could be determined based on monocultures though at the cost of longer run time

**Comment by hallucigenia-sparsa** [comment]
all this still assumes that uncertain events are evenly distributed across species, which may not be the case; so thresholding on uncertainty needs to remain optional.

**Comment by hariszaf** [comment]
Ok, I see. In the same notion, I guess we could also try to come up with class-wise thresholds; but maybe it's an overkill. By the way, here's the [paper](https://doi.org/10.1007/s41060-024-00584-z) I was looking at; it has some other ways for thresholds as well.

For now, I think we could do this: 

- [ ] have a default of 0.5 of the max entropy; so the prediction we keep are relatively more confident because their probability distribution is more skewed toward a particular class.
- [ ] implement the uncertainty thresholding for data-specific parameterization

What do you think ?

**Comment by hallucigenia-sparsa** [comment]
the simplest, given how many other projects are going on :). the drawback of the default threshold is that it becomes more stringent with more species, something we can inform users about in the GUI/yaml

**Comment by hariszaf** [comment]
So, I did this: 

- on the model (or best model if folds are being used) built, I get the `y_pred_classes` and `y_true_classes` 
- I then iterate over 10 quantiles of the maximum entropy and calculate its corresponding `classification_report`
- finally, I keep the the corresponding threshold of the percentile with the highest accuracy.

[Here](https://jsonblob.com/1332266553135259648) is how the dictionary with the reports looks like, and [here](https://github.com/hariszaf/CellScanner/commit/49af47d69e8c8a1a4cf0c933d6b474f7deb9315d#diff-3a4b7c803e19d3df60c0fc0caefadcba04eadadc8f22580e4a74a8476ca5a8c7R319) is the related function.

**Comment by hallucigenia-sparsa** [comment]
Nice. In your test case, having a more stringent threshold increased the accuracy, so having a threshold on the uncertainties is useful.

**Comment by hallucigenia-sparsa** [comment]
Is the threshold already supported in the yaml? Then I'll re-run my benchmarks

**Comment by hariszaf** [comment]
Give me some time to make some changes so we keep events to be dropped as `Unknown` and open a new PR with all the current changes.

**Comment by hariszaf** [comment]
@hallucigenia-sparsa I made a `dev` branch on the repo to be easier to test things before merging them to the actual running version. 

You may run 
```
git pull origin dev
```
to test the added feature.

**Comment by hallucigenia-sparsa** [comment]
I encountered 2 errors when trying to run the test case via the GUI with the latest version of CellScanner:
Error during prediction: TrainModelPanel' object has no attribute 'cs_uncertainty_threshold'
Error during prediction: could not convert string to float: ''

**Comment by hariszaf** [comment]
Gating now using thresholds on the raw data - before any kind of transformation. 
Gating is also enabled for the model training step. 

I close this issue and if something is not working as it should, or any other feature on gating is needed, we open a new one.
